python/ray/serve/_private/router.py

```diff
@@ -385,6 +385,42 @@ def update_running_replicas(self, running_replicas: List[RunningReplicaInfo]):
         """Shim for compatibility with the existing round robin scheduler."""
         return self.update_replicas([ActorReplicaWrapper(r) for r in running_replicas])
 
+    def _check_valid_candidate(self, replica_id: str) -> bool:
+        """Checks if the replica is a valid candidate based on its code version."""
+        internal_replica_context = ray.serve.context.get_internal_replica_context()
+        if (
+            internal_replica_context is None
+            or internal_replica_context.replica_tag is None
+        ):
+            return True
+
+        current_app = internal_replica_context.app_name
+        if not current_app:
+            return True
+
+        # NOTE(zcin): This is a slightly hacky way of getting the application
+        # of the deployment to which this router sends requests. We include
+        # application info in RunningReplicaInfo, and fetch it from there.
+        # A better solution would be to separate deployment names from app
+        # names and have Serve handles take a deployment name and application
+        # name as input upon initialization.
+        target_app = {
+            replica._replica_info.app_name for replica in self._replicas.values()
+        }.pop()
```

Comment:
IMO the best solution for this is to separate deployment name and application name, then initialize handles with `RayServeHandle(deployment_name, app_name)`. However this requires replacing with (app_name, deployment_name) wherever deployment name is used as an identifier, and that is a pretty big change. Before we make that change, this is a hacky-ish solution for making sure this change doesn't restrict sending requests between applications. LMK if there are any other suggestions! (for either the temporary solution or the more-permanent solution)

Reply:
I'm okay with this hack for now. Out of curiosity, is there any reason why we're using a set comprehension instead of picking the first replica in `self._replicas` and using its app name?